stm32u5: fix incorrect flas write width

vaussard opened the pull request:
This fixes #60724 and it was tested on the nucleo_u575zi_q using the test program mentioned in the ticket.

I loosely inspired myself from `write_qword()` in drivers/flash/flash_stm32wbax.c. But I see now that `write_ndwords()` from drivers/flash/flash_stm32h7x.c is maybe a better example.

Other drivers are hardcoding the write block size in `flash_stm32_valid_range()`, we could consider giving them the same treatment:
```
drivers/flash/flash_stm32f1x.c:173:     return (!write || (offset % 2 == 0 && len % 2 == 0U)) &&  
drivers/flash/flash_stm32g0x.c:53:      return (!write || (offset % 8 == 0 && len % 8 == 0)) &&  
drivers/flash/flash_stm32g4x.c:45:      return (!write || (offset % 8 == 0 && len % 8 == 0U)) &&  
drivers/flash/flash_stm32l4x.c:41:      return (!write || (offset % 8 == 0 && len % 8 == 0U)) &&  
drivers/flash/flash_stm32wbax.c:116:    return (!write || (offset % 16 == 0 && len % 16 == 0U)) &&  
drivers/flash/flash_stm32wbx.c:36:      return (!write || (offset % 8 == 0 && len % 8 == 0U)) &&
```

vaussard commented:
Note sure why twister fails in tests/subsys/pm/device_driver_init/ ?
```
INFO    - 614/614 qemu_cortex_m3            tests/subsys/pm/device_driver_init/pm.device_driver_init.pm_device_runtime  FAILED Failed (qemu 2.018s)
INFO    - /__w/zephyr/zephyr/twister-out/qemu_cortex_m3/tests/subsys/pm/device_driver_init/pm.device_driver_init.pm_device_runtime/handler.log
ERROR   - I: test_reg is now ON
I: test_reg_chained is now ON
*** Booting Zephyr OS build zephyr-v3.4.0-1554-g61873fab2882 ***
Running TESTSUITE device_driver_init
===================================================================
START - test_demo

    Assertion failed at WEST_TOPDIR/zephyr/tests/subsys/pm/device_driver_init/src/main.c:38: device_driver_init_test_demo: (gpio_config not equal to ((1U << 17) | (1U << 19)))
Unexpected config
 FAIL - test_demo in 0.001 seconds
===================================================================
TESTSUITE device_driver_init failed.
```

erwango commented:
> device_driver_init

@JordanYates Does this ring any bell ?

erwango commented on drivers/flash/flash_stm32l5x.c:
```diff
@@ -158,11 +158,13 @@ bool flash_stm32_valid_range(const struct device *dev, off_t offset,
 	volatile uint32_t *flash = (uint32_t *)(offset
 						+ CONFIG_FLASH_BASE_ADDRESS);
+	bool full_zero;
+	unsigned int i;
```
Why not int ?

erwango commented on drivers/flash/flash_stm32l5x.c:
```diff
@@ -178,16 +180,27 @@ static int write_dword(const struct device *dev, off_t offset, uint64_t val)
+		}
+	}
+	if (full_zero) {
+		for (unsigned int i = 0; i < FLASH_STM32_WRITE_BLOCK_SIZE; i++) {
```
```suggestion
		for (i = 0; i < FLASH_STM32_WRITE_BLOCK_SIZE; i++) {
```

erwango commented on drivers/flash/flash_stm32l5x.c:
```diff
@@ -197,8 +210,9 @@ static int write_dword(const struct device *dev, off_t offset, uint64_t val)
 	/* Perform the data write operation at the desired memory address */
-	flash[0] = (uint32_t)val;
-	flash[1] = (uint32_t)(val >> 32);
+	for (unsigned int i = 0; i < FLASH_STM32_WRITE_BLOCK_SIZE; i++) {
```
```suggestion
	for (i = 0; i < FLASH_STM32_WRITE_BLOCK_SIZE; i++) {
```

erwango commented on drivers/flash/flash_stm32l5x.c:
```diff
@@ -178,16 +180,27 @@ static int write_dword(const struct device *dev, off_t offset, uint64_t val)
-	     flash[1] != 0xFFFFFFFFUL) && val != 0UL) {
-		LOG_ERR("Word at offs %ld not erased", (long)offset);
-		return -EIO;
+	full_zero = true;
```
Can  you initialize it at declaration ?

JordanYates commented:
> > device_driver_init
> 
> @JordanYates Does this ring any bell ?

Yes this was caused by a combination of premature merging of #60411 and the changes introduced by #60458.
Reverting that PR will fix this issue.

erwango commented on drivers/flash/flash_stm32l5x.c:
```diff
@@ -178,16 +180,27 @@ static int write_dword(const struct device *dev, off_t offset, uint64_t val)
+			break;
+		}
+	}
+	if (full_zero) {
```
If I'm not mistaking, you're reverting the logic here.
It would be clearer to check:
- first: is the target zone not erased
- second: if not erased, verify that value is full_zeroes, error out if not the case.

vaussard replied:
Indeed it is reversed, but this is on purpose: the flash access is presumably much slower than a RAM access. So it is faster to check if the value is zero, and only then check the flash in the rare event that we write 0x00(...)00. In most cases, we can completely skip reading the flash. I can add a comment to make it more clear.

erwango replied:
ok, then you should test `if (!full_zero) {`

vaussard replied:
oh sure, sorry for this bug

erwango reviewed (changes requested):
Thanks for this
Some comments, and ideally as you mentioned `flash_stm32_valid_range` couldbe extended to most series. What about defining as weak  in flash_stm32.h? so it could be used as is unless a specific function is added when required

vaussard commented:
> What about defining as weak in flash_stm32.h?

Sure this sounds good. Why not in flash_stm32.c ?

vaussard committed:
drivers: flash: stm32l5: use write-block-size when writing

STM32L5 have a write block size of 8, but STM32U5 and STM32H5 have a
write block size of 16. Convert write_dword() and
flash_stm32_write_range() to write write-block-size data at a time.

Fixes #60724

Signed-off-by: Florian Vaussard <florian.vaussard@gmail.com>

vaussard committed:
drivers: flash: stm32l5: use write-block-size when validating

STM32L5 have a write block size of 8, but STM32U5 and STM32H5 have a
write block size of 16. Use write-block-size from the device tree
instead of hardcoding this value when validating the range of write
operations.

Fixes #60724

Signed-off-by: Florian Vaussard <florian.vaussard@gmail.com>

vaussard committed:
drivers: flash: stm32: add a weak flash_stm32_valid_range()

Most implementations have the same logic, with only a different write
block size. Now that we are using write-block-size from the device tree,
it is possible to use a default implementation that can be overridden if
necessary.

Signed-off-by: Florian Vaussard <florian.vaussard@gmail.com>

vaussard commented:
@erwango updated according to your suggestions. Ready for another review. Thanks!

erwango reviewed (approved):
That looks great.
Thanks @vaussard. Approving but blocking until we manage to perform HW tests.

erwango commented:
Tested ok

vaussard commented:
Many thanks @erwango.
